PulseAudio support

aykevl opened the pull request:
This PR adds PulseAudio protocol support. PulseAudio and PipeWire are both commonly used on desktop Linux systems, and because PipeWire is also typically used with the PulseAudio protocol this addition supports both.

(It's true that both also support the ALSA protocol, but the PulseAudio protocol integrates a little bit better).

The direct reason why I wrote this was because the ALSA backend wasn't working well for me with my built-in speakers (it was stuttering, no idea why) and wanted to see whether using PulseAudio would avoid that. And indeed it does. But I figured this might be useful for more people.

devgianlu commented:
I'll get to this later this week because I want to do a release soon-ish. Thank you for all the pull requests, you are doing a hell of a job!

aykevl commented:
Thanks!
I might fix some more issues with PulseAudio support in the meantime. For example, the package has some race conditions that I'd like to fix and there's no way to report the current audio volume back to Spotify.

(Also, there's some popping noises when starting/stopping playback, but that also happens for the alsa backend. I suspect it's because the audio output is just cut off instead of doing a very short fade out as I think is usually done in media players).

aykevl commented:
> (Also, there's some popping noises when starting/stopping playback, but that also happens for the alsa backend. I suspect it's because the audio output is just cut off instead of doing a very short fade out as I think is usually done in media players).

Nevermind, that also happens in YouTube on Firefox, with all 3 outputs I tested (desk speakers, laptop speakers, Bluetooth headphones). So maybe it's a PipeWire issue.

aykevl commented:
Updated after https://github.com/jfreymuth/pulse/pull/34 got merged. Improvements: less usage of raw PulseAudio commands, and using the PulseAudio volume on startup instead of defaulting to 100% (which can mean the maximum supported by the hardware on some systems, which is usually way too loud).

Once https://github.com/jfreymuth/pulse/pull/35 is in, I can update the PR to make sure external volume changes (through the system mixer) propagate back to Spotify. But that can also happen in a separate PR.

aykevl commented:
> The direct reason why I wrote this was because the ALSA backend wasn't working well for me with my built-in speakers (it was stuttering, no idea why)

I think I figured it out. The ALSA period size is set to 220, which meant that `snd_pcm_sw_params_set_avail_min` got 220, which would mean about 200 wakeups per second. No wonder it missed some of those.
I don't know why this period size is so small on my system (using PipeWire), but setting a higher value in `snd_pcm_sw_params_set_avail_min` (e.g. `bufferSize/4`) makes the stuttering go away.

aykevl commented:
Updated this PR:

 * Updated to include external volume updates (through a system mixer application) after https://github.com/jfreymuth/pulse/pull/35 got merged.
 * To make that possible in a reasonable way, I've refactored how volume updates are handled. This fixes a bug where frequent volume updates from a Spotify Client would be slown down or result in 429 HTTP errors (we can't send as many player status updates as Spotify sends us volume updates).
 * Fixed a bug where I unintentionally used linear volume instead of the default logarithmic (actually: cubic) volumes used by PulseAudio internally.

devgianlu commented:
@aykevl What's the state of this? Do you think it's ready after a rebase?

aykevl commented:
I'd say yes. I've been running go-librespot with this PR for a while now, it works fine. One benefit is that volume changes are applied immediately (well, except when Spotify itself is slow) instead of being delayed by ~500ms as with the ALSA backend.

There's one issue that I found: the seek location is not always accurate. Not a huge issue in my opinion and I can fix it in a separate PR, but if you prefer to have it fixed before merging I can look into it.

devgianlu commented:
> I'd say yes. I've been running go-librespot with this PR for a while now, it works fine.

Nice!

> There's one issue that I found: the seek location is not always accurate. Not a huge issue in my opinion and I can fix it in a separate PR, but if you prefer to have it fixed before merging I can look into it.

We can fix it later, no rush.

devgianlu commented on output/driver-alsa.go:
All files in the project use `_` insteaf of `-`, can you change it? For `mixer-alsa.go` as well.

aykevl replied:
`_` has a special meaning in Go: it is used for build tags (e.g. `somefile_linux.go`). I'm hesitant to use `_` because in the (in this case rather unlikely) case the suffix will be added to the list of build tags and the file suddenly doesn't compile anymore.

I can change it of course if you prefer `_`.

devgianlu replied:
Never thought about this. It's probably worth keeping it with `-` then (and rename the other files in the repository?).

aykevl commented:
Rebased, and did a very quick test that it still works.
Also updated the pulse package dependency: it now has a `volume.Norm()` function that does basically the same math as I previously did manually in this PR.
